<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Coordinate-based Genomic Visualization Package for R</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Coordinate-based Genomic Visualization Package for R
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;BentoBox&rsquo; version 0.1.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="BentoBox.html">BentoBox-package</a></td>
<td>BentoBox: Coordinate-based Genomic Visualization Package for R</td></tr>
<tr><td style="width: 25%;"><a href="bb_annoGenomeLabel.html">bb_annoGenomeLabel</a></td>
<td>Annotate genomic coordinates along the x or y-axis of a BentoBox plot</td></tr>
<tr><td style="width: 25%;"><a href="bb_annoHeatmapLegend.html">bb_annoHeatmapLegend</a></td>
<td>Add a color scale legend for heatmap-style plots</td></tr>
<tr><td style="width: 25%;"><a href="bb_annoHighlight.html">bb_annoHighlight</a></td>
<td>Annotates a highlight box around a specified genomic region of a BentoBox plot</td></tr>
<tr><td style="width: 25%;"><a href="bb_annoPixels.html">bb_annoPixels</a></td>
<td>Annotate pixels in a Hi-C plot</td></tr>
<tr><td style="width: 25%;"><a href="bb_annoSegments.html">bb_annoSegments</a></td>
<td>Annotates a line segment within a BentoBox plot</td></tr>
<tr><td style="width: 25%;"><a href="bb_annoText.html">bb_annoText</a></td>
<td>Annotates text within a BentoBox plot</td></tr>
<tr><td style="width: 25%;"><a href="bb_annoXaxis.html">bb_annoXaxis</a></td>
<td>Add an x-axis to a plot</td></tr>
<tr><td style="width: 25%;"><a href="bb_annoYaxis.html">bb_annoYaxis</a></td>
<td>Add a y-axis to a plot</td></tr>
<tr><td style="width: 25%;"><a href="bb_annoZoomLines.html">bb_annoZoomLines</a></td>
<td>Annotates zoom lines for a specified genomic region of a BentoBox plot</td></tr>
<tr><td style="width: 25%;"><a href="bb_assembly.html">bb_assembly</a></td>
<td>Make a 'bb_assembly' object for alternate TxDb, OrgDb,and BSgenome genomic annotation packages</td></tr>
<tr><td style="width: 25%;"><a href="bb_bedData.html">bb_bedData</a></td>
<td>BentoBox example BED data</td></tr>
<tr><td style="width: 25%;"><a href="bb_bedpeData.html">bb_bedpeData</a></td>
<td>BentoBox example BEDPE data</td></tr>
<tr><td style="width: 25%;"><a href="bb_CasesNYFL.html">bb_CasesNYFL</a></td>
<td>BentoBox example data for COVID-19 cases in New York and Florida</td></tr>
<tr><td style="width: 25%;"><a href="bb_CasesUSA.html">bb_CasesUSA</a></td>
<td>BentoBox example data for COVID-19 cases in the United States</td></tr>
<tr><td style="width: 25%;"><a href="bb_defaultPackages.html">bb_defaultPackages</a></td>
<td>Display the default genomic annotation packages associated with a genome build</td></tr>
<tr><td style="width: 25%;"><a href="bb_genomes.html">bb_genomes</a></td>
<td>Display the included available default genome assemblies</td></tr>
<tr><td style="width: 25%;"><a href="bb_gmCTCFData.html">bb_gmCTCFData</a></td>
<td>BentoBox example GM12878 CTCF signal data</td></tr>
<tr><td style="width: 25%;"><a href="bb_gmH3K27acData.html">bb_gmH3K27acData</a></td>
<td>BentoBox example GM12878 H3K27ac signal data</td></tr>
<tr><td style="width: 25%;"><a href="bb_gmHicData.html">bb_gmHicData</a></td>
<td>BentoBox example GM12878 Hi-C data</td></tr>
<tr><td style="width: 25%;"><a href="bb_gwasData.html">bb_gwasData</a></td>
<td>BentoBox example GWAS data</td></tr>
<tr><td style="width: 25%;"><a href="bb_imrCTCFData.html">bb_imrCTCFData</a></td>
<td>BentoBox example IMR90 CTCF signal data</td></tr>
<tr><td style="width: 25%;"><a href="bb_imrH3K27acData.html">bb_imrH3K27acData</a></td>
<td>BentoBox example IMR90 H3K27ac signal data</td></tr>
<tr><td style="width: 25%;"><a href="bb_imrHicData.html">bb_imrHicData</a></td>
<td>BentoBox example IMR90 Hi-C data</td></tr>
<tr><td style="width: 25%;"><a href="bb_pageCreate.html">bb_pageCreate</a></td>
<td>Create a page for a BentoBox layout</td></tr>
<tr><td style="width: 25%;"><a href="bb_pageGuideHide.html">bb_pageGuideHide</a></td>
<td>Remove guides from a BentoBox page</td></tr>
<tr><td style="width: 25%;"><a href="bb_pageGuideHorizontal.html">bb_pageGuideHorizontal</a></td>
<td>Draw a horizontal guideline at a specified y-coordinate on a BentoBox page</td></tr>
<tr><td style="width: 25%;"><a href="bb_pageGuideShow.html">bb_pageGuideShow</a></td>
<td>Reshow guides drawn with 'bb_pageCreate', 'bb_pageGuideHorizontal', and 'bb_pageGuideVertical'</td></tr>
<tr><td style="width: 25%;"><a href="bb_pageGuideVertical.html">bb_pageGuideVertical</a></td>
<td>Draw a vertical guideline at a specified x-coordinate on a BentoBox page</td></tr>
<tr><td style="width: 25%;"><a href="bb_pagePlotPlace.html">bb_pagePlotPlace</a></td>
<td>Place a BentoBox plot that has been previously created but not drawn</td></tr>
<tr><td style="width: 25%;"><a href="bb_pagePlotRemove.html">bb_pagePlotRemove</a></td>
<td>Remove BentoBox plots and annotations</td></tr>
<tr><td style="width: 25%;"><a href="bb_params.html">bb_params</a></td>
<td>bb_params: BentoBox parameters object</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotBase.html">bb_plotBase</a></td>
<td>Plot a base R plot in a BentoBox layout</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotCircle.html">bb_plotCircle</a></td>
<td>Plot a circle within a BentoBox layout</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotGenes.html">bb_plotGenes</a></td>
<td>Plot a gene track for a specified genomic region</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotGenomeLabel.html">bb_plotGenomeLabel</a></td>
<td>Plot genomic coordinates along the x or y-axis of a BentoBox plot</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotGG.html">bb_plotGG</a></td>
<td>Plot a ggplot2 plot, gtable, or grob object in a BentoBox layout</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotHicRectangle.html">bb_plotHicRectangle</a></td>
<td>Plot a triangular Hi-C interaction matrix in a rectangular format</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotHicSquare.html">bb_plotHicSquare</a></td>
<td>Plot a Hi-C interaction matrix in a square format</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotHicTriangle.html">bb_plotHicTriangle</a></td>
<td>Plot a Hi-C interaction matrix in a triangular format</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotIdeogram.html">bb_plotIdeogram</a></td>
<td>Plot a chromosome ideogram with or without cytobands</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotLegend.html">bb_plotLegend</a></td>
<td>Plot a legend</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotManhattan.html">bb_plotManhattan</a></td>
<td>Plot a Manhattan plot</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotPairs.html">bb_plotPairs</a></td>
<td>Plot paired-end genomic range elements</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotPairsArches.html">bb_plotPairsArches</a></td>
<td>Plot paired-end genomic range data in an arch style</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotPolygon.html">bb_plotPolygon</a></td>
<td>Plot a polygon within a BentoBox layout</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotRanges.html">bb_plotRanges</a></td>
<td>Plot genomic range elements in a pileup or collapsed format</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotRaster.html">bb_plotRaster</a></td>
<td>Plot a raster object within a BentoBox layout</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotRect.html">bb_plotRect</a></td>
<td>Plot a rectangle within a BentoBox layout</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotSegments.html">bb_plotSegments</a></td>
<td>Draw a line segment within a BentoBox layout</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotSignal.html">bb_plotSignal</a></td>
<td>Plot any kind of signal track data for a single chromosome</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotText.html">bb_plotText</a></td>
<td>Plot text within a BentoBox layout</td></tr>
<tr><td style="width: 25%;"><a href="bb_plotTranscripts.html">bb_plotTranscripts</a></td>
<td>Plot gene transcripts in a pileup style for a single chromosome</td></tr>
<tr><td style="width: 25%;"><a href="bb_readBigwig.html">bb_readBigwig</a></td>
<td>Read a bigWig file and return it as a data frame</td></tr>
<tr><td style="width: 25%;"><a href="bb_readHic.html">bb_readHic</a></td>
<td>Read a .hic file and return Hi-C data as a dataframe</td></tr>
<tr><td style="width: 25%;"><a href="bb_VaccinesNYFL.html">bb_VaccinesNYFL</a></td>
<td>BentoBox example data for COVID-19 vaccinations in New York and Florida</td></tr>
<tr><td style="width: 25%;"><a href="BentoBox.html">BentoBox</a></td>
<td>BentoBox: Coordinate-based Genomic Visualization Package for R</td></tr>
<tr><td style="width: 25%;"><a href="c.html">c</a></td>
<td>Combine multiple bb_params objects into a vector</td></tr>
<tr><td style="width: 25%;"><a href="colorby.html">colorby</a></td>
<td>Handle BentoBox color scaling parameters</td></tr>
<tr><td style="width: 25%;"><a href="cytoBand.Dmelanogaster.UCSC.dm6.html">cytoBand.Dmelanogaster.UCSC.dm6</a></td>
<td>UCSC CytoBand information for the dm6 genome assembly</td></tr>
<tr><td style="width: 25%;"><a href="cytoBand.Drerio.UCSC.danRer10.html">cytoBand.Drerio.UCSC.danRer10</a></td>
<td>UCSC CytoBand information for the danRer10 genome assembly</td></tr>
<tr><td style="width: 25%;"><a href="cytoBand.Hsapiens.UCSC.hg18.html">cytoBand.Hsapiens.UCSC.hg18</a></td>
<td>UCSC CytoBand information for the hg18 genome assembly</td></tr>
<tr><td style="width: 25%;"><a href="cytoBand.Hsapiens.UCSC.hg19.html">cytoBand.Hsapiens.UCSC.hg19</a></td>
<td>UCSC CytoBand information for the hg19 genome assembly</td></tr>
<tr><td style="width: 25%;"><a href="cytoBand.Hsapiens.UCSC.hg38.html">cytoBand.Hsapiens.UCSC.hg38</a></td>
<td>UCSC CytoBand information for the hg38 genome assembly</td></tr>
<tr><td style="width: 25%;"><a href="cytoBand.Mmusculus.UCSC.mm10.html">cytoBand.Mmusculus.UCSC.mm10</a></td>
<td>UCSC CytoBand information for the mm10 genome assembly</td></tr>
<tr><td style="width: 25%;"><a href="cytoBand.Mmusculus.UCSC.mm9.html">cytoBand.Mmusculus.UCSC.mm9</a></td>
<td>UCSC CytoBand information for the mm9 genome assembly</td></tr>
<tr><td style="width: 25%;"><a href="cytoBand.Rnorvegicus.UCSC.rn5.html">cytoBand.Rnorvegicus.UCSC.rn5</a></td>
<td>UCSC CytoBand information for the rn5 genome assembly</td></tr>
<tr><td style="width: 25%;"><a href="cytoBand.Rnorvegicus.UCSC.rn6.html">cytoBand.Rnorvegicus.UCSC.rn6</a></td>
<td>UCSC CytoBand information for the rn6 genome assembly</td></tr>
</table>
</body></html>
