<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Coordinate-based Genomic Visualization Package for R • BentoBox</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Coordinate-based Genomic Visualization Package for R">
<meta property="og:description" content="&gt;
    BentoBox is a coordinate-based genomic visualization package for R. It grants users
    the ability to programmatically produce complex, multi-paneled figures. Tailored for
    genomics, BentoBox allows users to visualize large complex genomic datasets and
    provides exquisite control over how plots are placed and arranged on a page. ">
<meta property="og:image" content="https://phanstiellab.github.io/bentobox/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">BentoBox</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/our_philosophy.html">Our Philosophy</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/introduction_to_bentobox.html">Introduction to BentoBox</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PhanstielLab/BentoBox">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#-" class="anchor"></a><img src="reference/figures/bento-logotype-singleline-black.png" width="400px" style="background-color:white;border=transparent"><img src="reference/figures/logo.png" align="right" width="140px" style="padding-left:20px; background-color:white">
</h1></div>
<p><code>BentoBox</code> is a genomic data vizualization package for R. Using <code>grid</code> graphics, <code>BentoBox</code> empowers users to programmatically and flexibly generate multi-panel figures. <code>BentoBox</code> accomplishes these goals by utilizing 1) a coordinate-based plotting system, and 2) edge-to-edge containerized data visualization. The coordinate-based plotting system grants users precise control over the size, position, and arrangement of plots. Its edge-to-edge plotting functions preserve the mapping between user-specified containers and the represented data. This allows users to stack plots with confidence that vertically aligned data will correspond to the same regions. For more information about BenoBox’s philosophy and design, check out the <code>Our Philosophy</code> page.</p>
<p>Specialized for genomic data, <code>BentoBox</code> also contains functions to read and plot multi-omic data quickly and easily. <code>BentoBox</code> can address an endless number of use cases, including: dynamic exploration of genomic data, arrangment into multi-omic layouts, and survey plotting for quickly viewing data across the genome. Check out our <code>vignettes</code> for detailed examples and suggested use cases!</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>BentoBox can be installed from GitHub as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"remotes"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span>

<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"PhanstielLab/BentoBox"</span><span class="op">)</span>
<span class="fu">package.version</span><span class="op">(</span><span class="st">"BentoBox"</span><span class="op">)</span></code></pre></div>
</div>
<div id="examples" class="section level2">
<h2 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h2>
<div id="quick-plotting" class="section level3">
<h3 class="hasAnchor">
<a href="#quick-plotting" class="anchor"></a>Quick plotting</h3>
<p>We can use any of the plotting functions to quickly plot a single data type by simply ignoring the arguments that define the plotting location (i.e. <code>x</code>, <code>y</code>, <code>width</code>, <code>height</code>, <code>just</code>, <code>default.units</code>).</p>
<p>Lets demonstrate this by plotting the example Hi-C data included in bb_hicData (chr21:28000000-30300000):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load BentoBox</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/bentobox">BentoBox</a></span><span class="op">)</span>

<span class="co">## Load example Hi-C data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"bb_hicData"</span><span class="op">)</span>

<span class="co">## Quick plot Hi-C data</span>
<span class="va">hicPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bb_plotHicSquare.html">bb_plotHicSquare</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">bb_hicData</span>,
                            chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/quickHicExample-1.png" style="display: block; margin: auto;"></p>
<p>In addition to plotting the Hi-C Data, this creates an S3 object of class <code>bb_hicSquare</code> which stores all of the information about the region being plotted, the genome assembly, and any plotting parameters.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">hicPlot</span><span class="op">)</span>
<span class="co">## [1] "bb_hicSquare"</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">hicPlot</span><span class="op">)</span>
<span class="co">## $chrom</span>
<span class="co">## [1] "chr21"</span>
<span class="co">## </span>
<span class="co">## $chromstart</span>
<span class="co">## [1] 2.8e+07</span>
<span class="co">## </span>
<span class="co">## $chromend</span>
<span class="co">## [1] 30300000</span>
<span class="co">## </span>
<span class="co">## $altchrom</span>
<span class="co">## [1] "chr21"</span>
<span class="co">## </span>
<span class="co">## $altchromstart</span>
<span class="co">## [1] 2.8e+07</span>
<span class="co">## </span>
<span class="co">## $altchromend</span>
<span class="co">## [1] 30300000</span>
<span class="co">## </span>
<span class="co">## $assembly</span>
<span class="co">## $Genome</span>
<span class="co">## [1] "hg19"</span>
<span class="co">## </span>
<span class="co">## $TxDb</span>
<span class="co">## [1] "TxDb.Hsapiens.UCSC.hg19.knownGene"</span>
<span class="co">## </span>
<span class="co">## $OrgDb</span>
<span class="co">## [1] "org.Hs.eg.db"</span>
<span class="co">## </span>
<span class="co">## $gene.id.column</span>
<span class="co">## [1] "ENTREZID"</span>
<span class="co">## </span>
<span class="co">## $display.column</span>
<span class="co">## [1] "SYMBOL"</span>
<span class="co">## </span>
<span class="co">## $BSgenome</span>
<span class="co">## [1] "BSgenome.Hsapiens.UCSC.hg19"</span>
<span class="co">## </span>
<span class="co">## attr(,"class")</span>
<span class="co">## [1] "bb_assembly"</span>
<span class="co">## </span>
<span class="co">## $resolution</span>
<span class="co">## [1] 10000</span>
<span class="co">## </span>
<span class="co">## $x</span>
<span class="co">## NULL</span>
<span class="co">## </span>
<span class="co">## $y</span>
<span class="co">## NULL</span>
<span class="co">## </span>
<span class="co">## $width</span>
<span class="co">## NULL</span>
<span class="co">## </span>
<span class="co">## $height</span>
<span class="co">## NULL</span>
<span class="co">## </span>
<span class="co">## $just</span>
<span class="co">## [1] "left" "top" </span>
<span class="co">## </span>
<span class="co">## $color_palette</span>
<span class="co">## function (n) </span>
<span class="co">## {</span>
<span class="co">##     x &lt;- ramp(seq.int(0, 1, length.out = n))</span>
<span class="co">##     if (ncol(x) == 4L) </span>
<span class="co">##         rgb(x[, 1L], x[, 2L], x[, 3L], x[, 4L], maxColorValue = 255)</span>
<span class="co">##     else rgb(x[, 1L], x[, 2L], x[, 3L], maxColorValue = 255)</span>
<span class="co">## }</span>
<span class="co">## &lt;bytecode: 0x7fd629903b70&gt;</span>
<span class="co">## &lt;environment: 0x7fd629901320&gt;</span>
<span class="co">## </span>
<span class="co">## $zrange</span>
<span class="co">## [1]  0 70</span>
<span class="co">## </span>
<span class="co">## $half</span>
<span class="co">## [1] "both"</span>
<span class="co">## </span>
<span class="co">## $grobs</span>
<span class="co">## gTree[GRID.gTree.1] </span>
<span class="co">## </span>
<span class="co">## attr(,"class")</span>
<span class="co">## [1] "bb_hicSquare"</span>
<span class="co">## attr(,"plotted")</span>
<span class="co">## [1] TRUE</span></code></pre></div>
<p><code>bb_hicSquare</code> objects, and all objects created by <code>bb_plot</code> functions, can saved and used later in multi-plot arrangements (see <code><a href="reference/bb_pagePlotPlace.html">bb_pagePlotPlace()</a></code> and related vignettes). For different types of plots check out the rest of the plotting functions in the Reference section.</p>
<p>We can use our own data by using <code>bb_read</code> functions. This can be useful if there are additional processing steps required.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/bb_readHic.html">bb_readHic</a></span><span class="op">(</span>hicFile <span class="op">=</span> <span class="st">"path/to/file.hic"</span>, <span class="va">...</span><span class="op">)</span></code></pre></div>
<p>Or if no processing is required, we can simply provide the file path directly to the <code>bb_plot</code> function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/bb_plotHicSquare.html">bb_plotHicSquare</a></span><span class="op">(</span>hicData <span class="op">=</span> <span class="st">"path/to/file.hic"</span>, <span class="va">...</span><span class="op">)</span></code></pre></div>
</div>
<div id="multi-plot-layouts" class="section level3">
<h3 class="hasAnchor">
<a href="#multi-plot-layouts" class="anchor"></a>Multi-plot layouts</h3>
<p>To add annotation features to a <code>bb_plot</code> object, or to view multiple <code>bb_plot</code> objects simultaneously, we must:</p>
<ol>
<li>Create a BentoBox coordinate page with <code><a href="reference/bb_pageCreate.html">bb_pageCreate()</a></code>.<br>
</li>
<li>Provide values for the placement arguments (<code>x</code>, <code>y</code>, <code>width</code>, <code>height</code>, <code>just</code>, <code>default.units</code>) in <code>bb_plot</code> functions.</li>
</ol>
<p>As an annotation example, lets say we want to add a color scale legend for a Hi-C plot:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load BentoBox</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/bentobox">BentoBox</a></span><span class="op">)</span>

<span class="co">## Load example Hi-C data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"bb_hicData"</span><span class="op">)</span>

<span class="co">## Create a BentoBox page</span>
<span class="fu"><a href="reference/bb_pageCreate.html">bb_pageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">3.25</span>, height <span class="op">=</span> <span class="fl">3.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span><span class="op">)</span>

<span class="co">## Plot Hi-C data with placing information</span>
<span class="va">hicPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bb_plotHicSquare.html">bb_plotHicSquare</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">bb_hicData</span>,
                            chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
                            x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">2.5</span>, height <span class="op">=</span> <span class="fl">2.5</span>, default.units <span class="op">=</span> <span class="st">"inches"</span><span class="op">)</span>

<span class="co">## Add color scale annotation</span>
<span class="fu"><a href="reference/bb_annoHeatmapLegend.html">bb_annoHeatmapLegend</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">hicPlot</span>,
                     x <span class="op">=</span> <span class="fl">2.85</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">0.1</span>, height <span class="op">=</span> <span class="fl">1.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-5-1.png" style="display: block; margin: auto;"></p>
<p>By default, all placement coordinates are relative to the top right corner of the page/plot. This can be adjusted for each function individually with the <code>just</code> parameter. This allows us to use the most convenient coordinates to place our plot and the ability to right, left, center, top, bottom, or middle align every plotting element. For more information on plot placement, check out the vignettes.</p>
<p>Now, let’s suppose we have both Hi-C data and signal track data such as H3K27Ac bigWig files. We could enter the region information (chr21:28000000-30300000) into both functions. But we can streamline this process by taking advantage of the <code><a href="reference/bb_params.html">bb_params()</a></code> function. Since many functions share arguments, <code><a href="reference/bb_params.html">bb_params()</a></code> lets us reduce repetition by defining shared arguments in one place. Furthermore, we can concatenate these <code>bb_params</code> objects, to create useful combinations of parameter settings. For more information about <code><a href="reference/bb_params.html">bb_params()</a></code> usage, refer to the documentation or vignettes. The following code shows how we can utilize these functions to plot Hi-C data H3K27Ac signal tracks, and add a genomic label to a shared region:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load BentoBox</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/bentobox">BentoBox</a></span><span class="op">)</span>

<span class="co">## Load example Hi-C and signal track (H3K27Ac) data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"bb_hicData"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"bb_signalData"</span><span class="op">)</span>

<span class="co">## Define common region &amp; plot-placement parameters</span>
<span class="va">region</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bb_params.html">bb_params</a></span><span class="op">(</span>chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span><span class="op">)</span>
<span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bb_params.html">bb_params</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">2.5</span>, default.units <span class="op">=</span> <span class="st">"inches"</span><span class="op">)</span>

<span class="co">## Create a BentoBox page</span>
<span class="fu"><a href="reference/bb_pageCreate.html">bb_pageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">3.25</span>, height <span class="op">=</span> <span class="fl">4</span>, default.units <span class="op">=</span> <span class="st">"inches"</span><span class="op">)</span>

<span class="co">## Plot Hi-C data with combined bb_params objects</span>
<span class="va">hicPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bb_plotHicSquare.html">bb_plotHicSquare</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">bb_hicData</span>,
                            params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">region</span>, <span class="va">params</span><span class="op">)</span>,
                            y <span class="op">=</span> <span class="fl">0.25</span>, height <span class="op">=</span> <span class="va">params</span><span class="op">$</span><span class="va">width</span><span class="op">)</span>

<span class="co">## Add color scale annotation</span>
<span class="fu"><a href="reference/bb_annoHeatmapLegend.html">bb_annoHeatmapLegend</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">hicPlot</span>,
                     x <span class="op">=</span> <span class="fl">2.85</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">0.1</span>, height <span class="op">=</span> <span class="fl">1.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span><span class="op">)</span>

<span class="co">## Plot H3K27Ac signal track</span>
<span class="fu"><a href="reference/bb_plotSignal.html">bb_plotSignal</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">bb_signalData</span>,
              params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">region</span>, <span class="va">params</span><span class="op">)</span>,
              y <span class="op">=</span> <span class="fl">3</span>, height <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span>

<span class="co">## Label genome coordinates</span>
<span class="fu"><a href="reference/bb_plotGenomeLabel.html">bb_plotGenomeLabel</a></span><span class="op">(</span>params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">region</span>, <span class="va">params</span><span class="op">)</span>,
                   y <span class="op">=</span> <span class="fl">3.5</span>, length <span class="op">=</span> <span class="va">params</span><span class="op">$</span><span class="va">width</span>, scale <span class="op">=</span> <span class="st">"bp"</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-6-1.png" style="display: block; margin: auto;"></p>
<p>You might notice that using <code><a href="reference/bb_params.html">bb_params()</a></code> we can remove quite a few of the plotting arguments. The only things we need to specify are the data to use and the y-coordinate to place each plot. Using <code><a href="https://rdrr.io/r/base/c.html">c()</a></code> we can also combine distinct sets of <code><a href="reference/bb_params.html">bb_params()</a></code> to create different combinations of paramters for each function. We can also override arguments in <code><a href="reference/bb_params.html">bb_params()</a></code> objects by defining them in local functions if needed.</p>
<p>Another useful feature of <code><a href="reference/bb_params.html">bb_params()</a></code> is the ability to build regions and plots from gene names, given genomic annotation packages are loaded. You can check which data packages belong to which assemblies by using <code><a href="reference/bb_defaultPackages.html">bb_defaultPackages(Genome)</a></code> or make your own with <code><a href="reference/bb_assembly.html">bb_assembly()</a></code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/bb_defaultPackages.html">bb_defaultPackages</a></span><span class="op">(</span><span class="st">"hg19"</span><span class="op">)</span>
<span class="co">## List of 6</span>
<span class="co">##  $ Genome        : chr "hg19"</span>
<span class="co">##  $ TxDb          : chr "TxDb.Hsapiens.UCSC.hg19.knownGene"</span>
<span class="co">##  $ OrgDb         : chr "org.Hs.eg.db"</span>
<span class="co">##  $ gene.id.column: chr "ENTREZID"</span>
<span class="co">##  $ display.column: chr "SYMBOL"</span>
<span class="co">##  $ BSgenome      : chr "BSgenome.Hsapiens.UCSC.hg19"</span>
<span class="co">##  - attr(*, "class")= chr "bb_assembly"</span></code></pre></div>
<p>Here is an example of using <code><a href="reference/bb_params.html">bb_params()</a></code> to plot the IL1B gene using the <code>gene</code> and <code>geneBuffer</code> arguments:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load assembly packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span>

<span class="co">## Plot IL1B with no buffer</span>
<span class="fu"><a href="reference/bb_plotGenes.html">bb_plotGenes</a></span><span class="op">(</span>params <span class="op">=</span> <span class="fu"><a href="reference/bb_params.html">bb_params</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="st">'IL1B'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-8-1.png" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Plot genes with a 1000000bp (1Mb) buffer, centered on IL1B</span>
<span class="fu"><a href="reference/bb_plotGenes.html">bb_plotGenes</a></span><span class="op">(</span>params <span class="op">=</span> <span class="fu"><a href="reference/bb_params.html">bb_params</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="st">'IL1B'</span>, geneBuffer <span class="op">=</span> <span class="fl">1000000</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-9-1.png" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="a-word-of-caution" class="section level2">
<h2 class="hasAnchor">
<a href="#a-word-of-caution" class="anchor"></a>A word of caution</h2>
<p><code>BentoBox</code> is incredibly flexible and functional. However, due to this flexibility and like all programming packages, it may not always prevent users from making unintentional mistakes. If plot sizes are entered incorrectly or data is mishandled, it is possible to connect multi-omic data incorrectly. Make sure you utilize package features that reduce human error and increase re-usability of code to get the most milage out of BentoBox.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/PhanstielLab/BentoBox/">https://​github.com/​PhanstielLab/​BentoBox/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;= 2)</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/nekramer">Nicole Kramer</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9617-9671" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/EricSDavis">Eric S. Davis</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-4051-3217" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/cwenger">Craig Wenger</a> <br><small class="roles"> Author </small>  </li>
<li>
<a href="https://github.com/dphansti">Douglas H. Phanstiel</a> <br><small class="roles"> Author, copyright holder </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/nekramer">Nicole Kramer</a>, <a href="https://github.com/EricSDavis">Eric S. Davis</a>, <a href="https://github.com/cwenger">Craig Wenger</a>, <a href="https://github.com/dphansti">Douglas H. Phanstiel</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
